### PR TITLE
RUMM-2239 Replace `Telemetry` injection with `DD.telemetry` global

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

🚚 This PR simplifies the way we expose `Telemetry` monitor to different parts of the codebase. 

Instead of passing `telemetry: Telemetry?` through initializers, we now use `DD.telemetry` as global util. It removes the need of passing the reference all around from `Datadog` facade to downstream components.

It also makes the [`Telemetry`](https://github.com/DataDog/dd-sdk-ios/blob/develop/Sources/Datadog/Core/DD/Telemetry.swift) close to available for use in cross-platform frameworks (now it only requires public API wrapper - to be added later // cc @fuzzybinary).

### How?

* Removed all `telemetry: Telemetry?` references passed through different initializers.
* Exposed `DD.telemetry` instead.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests
- [ ] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: Tests/DatadogTests/Datadog/Mocks/CoreMocks.swift
Comment (ncreated) [inline Tests/DatadogTests/Datadog/Mocks/CoreMocks.swift:1326]: Not so happy with having two different factory methods:
* `static func mockWith<CL: CoreLogger>(logger: CL) -> DDMock<CL, TelemetryMock> {`
* `static func mockWith<TM: Telemetry>(telemetry: TM) -> DDMock<CoreLoggerMock, TM>`

but I'd rather wait for another thing that we may add to `DD` global before finding a better shape for it.

File: Sources/Datadog/Datadog.swift
Comment (ncreated) [reply Sources/Datadog/Datadog.swift:221]: Good point 👍, I think we forgot about multiple instances of the SDK @maxep (well, AFAIR we discussed `core` <> `telemetry` relationship some time ago, but not recently when deciding on shift to `DD.telemetry`).

To some extent it isn't important which instance of the SDK uploads the telemetry from a running process. After all, telemetry will land in the same (Telemetry) Org and will consume the same user quota. Telemetry events by their nature are anonymous and self-contained, so I can't imagine situation where this would become problematic (any thoughts @fuzzybinary @maxep @xgouchet ?).

The only instance-specific information is the RUM context we inject into telemetry events:

https://github.com/DataDog/dd-sdk-ios/blob/f330ac00db328e110ca5287818c58cc9aaa706ec/Sources/Datadog/RUM/RUMTelemetry.swift#L74-L76

It will differ (or become wrong) based on initialization order if `DD.telemetry` is global.

Comment (ncreated) [reply Sources/Datadog/Datadog.swift:221]: Also, maybe this is something we should decide on later, when the codebase is closer to its final V2 shape. Now we're clearly lacking utilities and `core` reference availability in many places, hence the `telemetry: Telemetry?` propagation is pretty tedious. I can imagine this will be simpler in few sprints from now (right before splitting into modules). WDYT @maxep ?

Comment (maxep) [reply Sources/Datadog/Datadog.swift:221]: Indeed, we had a brief discussion and concluded that a telemetry event could be dispatched on any user org. The idea was to register multiple cores to the global telemetry and let it balance events on all of them. But indeed, we have `core`-specific event data, e.g. [RUM Context](https://github.com/DataDog/dd-sdk-ios/blob/develop/Sources/Datadog/RUM/RUMTelemetry.swift#L110L112) and [App ID](https://github.com/DataDog/dd-sdk-ios/blob/develop/Sources/Datadog/RUM/RUMTelemetry.swift#L117).
Injecting `core` everywhere can be problematic (retain circle), especially in low-levels (like data persistency and uploads) and I don't think we need RUM data on every telemetry events 🤔 so it could be on a case by case: either the telemetry uses a specific `core`, or it can pick a random one from the registry.

As for now, multiple calls to init will override the previous SDK instance (not only the telemetry). We don't yet support multiple instances of the SDK so I think we can move forward with the `DD.telemetry` and tackle that when we commit to multiple instance support. How does it sounds?

Comment (ncreated) [reply Sources/Datadog/Datadog.swift:221]: > We don't yet support multiple instances of the SDK so I think we can move forward with the DD.telemetry and tackle that when we commit to multiple instance support. How does it sounds?

Sounds good for me 👍. From the experience on changing V1 into V2 it seems very pragmatic to incur this small tech debt now and pay it off later when the codebase will be closer to its final shape. Especially, that we may need the same for `DD.logger` (it has instance-specific capability of `.verbosity` threshold), tackling it once at whole makes more sense.

File: Sources/Datadog/Core/DD/Telemetry.swift
Comment (ncreated) [reply Sources/Datadog/Core/DD/Telemetry.swift]: Sounds good, changed 👍. 

It would be also great to make it consistent among the entire codebase + enforced by the linter, so I created JIRA: _`RUMM-2340` Enforce consistent `NOP` prefix for all no-op implementations_

**Review by fuzzybinary — APPROVED** [review]

**Pull request merged**
